chain/stmgr/stmgr.go

```diff
@@ -755,6 +757,89 @@ func (sm *StateManager) ValidateChain(ctx context.Context, ts *types.TipSet) err
 	return nil
 }
 
-func (sm *StateManager) SetVMConstructor(nvm func(cid.Cid, abi.ChainEpoch, vm.Rand, blockstore.Blockstore, vm.SyscallBuilder) (*vm.VM, error)) {
+func (sm *StateManager) SetVMConstructor(nvm func(cid.Cid, abi.ChainEpoch, vm.Rand, blockstore.Blockstore, vm.SyscallBuilder, vm.VestedCalculator) (*vm.VM, error)) {
 	sm.newVM = nvm
 }
+
+type GenesisMsigEntry struct {
+	totalFunds abi.TokenAmount
+	unitVest   abi.TokenAmount
+}
+
+func (sm *StateManager) setupGenesisMsigs(ctx context.Context) error {
+	gb, err := sm.cs.GetGenesis()
+	if err != nil {
+		return xerrors.Errorf("getting genesis block: %w", err)
+	}
+
+	gts, err := types.NewTipSet([]*types.BlockHeader{gb})
+	if err != nil {
+		return xerrors.Errorf("getting genesis tipset: %w", err)
+	}
+
+	st, _, err := sm.TipSetState(ctx, gts)
+	if err != nil {
+		return xerrors.Errorf("getting genesis tipset state: %w", err)
+	}
+
+	r, err := adt.AsMap(sm.cs.Store(ctx), st)
+	if err != nil {
+		return xerrors.Errorf("getting genesis actors: %w", err)
+	}
+
+	totalsByEpoch := make(map[abi.ChainEpoch]abi.TokenAmount)
+	var act types.Actor
+	err = r.ForEach(&act, func(k string) error {
+		if act.Code == builtin.MultisigActorCodeID {
+			var s multisig.State
+			err := sm.cs.Store(ctx).Get(ctx, act.Head, &s)
+			if err != nil {
+				return err
+			}
+
+			if s.StartEpoch != 0 {
+				return xerrors.New("genesis multisig doesn't start vesting at epoch 0!")
+			}
+
+			ot, f := totalsByEpoch[s.UnlockDuration]
+			if f {
+				totalsByEpoch[s.UnlockDuration] = big.Add(ot, s.InitialBalance)
+			} else {
+				totalsByEpoch[s.UnlockDuration] = s.InitialBalance
+			}
+
+		}
+		return nil
+	})
+
+	if err != nil {
+		return xerrors.Errorf("error setting up composite genesis multisigs: %w", err)
+	}
+
+	sm.genesisMsigs = make([]multisig.State, 0, len(totalsByEpoch))
+	for k, v := range totalsByEpoch {
+		ns := multisig.State{
+			InitialBalance: v,
+			UnlockDuration: k,
+			PendingTxns:    cid.Undef,
+		}
+		sm.genesisMsigs = append(sm.genesisMsigs, ns)
+	}
+
+	return nil
+}
+
+func (sm *StateManager) GetVestedFunds(ctx context.Context, height abi.ChainEpoch) (abi.TokenAmount, error) {
+
+	if sm.genesisMsigs == nil {
+		err := sm.setupGenesisMsigs(ctx)
+		if err != nil {
+			return big.Zero(), xerrors.Errorf("failed to setup genesis msig entries: %w", err)
+		}
+	}
+	vf := big.Zero()
```

Comment:
`vf` is what is currently locked iiuc, VestedFunds should be `FIL_Alloc_Genesis - vf`.

Reply:
doh!